third_party/amd/lib/TritonAMDGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandWMMA.cpp

```diff
@@ -0,0 +1,226 @@
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#ifdef USE_ROCM
+
+#include "../PatternTritonGPUOpToLLVM.h"
+#include "SharedToDotOperandMatrixCore.h"
+#include "Utility.h"
+
+using ::mlir::triton::gpu::AMDWmmaEncodingAttr;
+using ::mlir::triton::gpu::DotOperandEncodingAttr;
+using ::mlir::triton::gpu::getOrder;
+using ::mlir::triton::gpu::getShapePerCTA;
+using ::mlir::triton::gpu::SharedEncodingAttr;
+
+namespace SharedToDotOperandWMMA {
+
+/**
+ * @brief This function maps particular load of wmma dot operand to element
+ * indexes(row, col)
+ *
+ * Whole tensor is broken into "blocks" of warps along "non-K" axis.
+ * One block could be processed by multiple warps.
+ * One warp works on a piece of tensor size elemsPerInstr[0] x K.
+ * Each of these pieces is broken into "tiles" of size elemsPerInstr[0] x
+ * elemsPerInstr[1].
+ *
+ * Total offset of element is a sum of following values:
+ * 1. Offset of warp block in tensor
+ * 2. Offset of warp inside one warp block
+ * 3. Offset of tile in one warp
+ * 4. Offset of one lane data in a tile
+ * 5. Offset of particular element of tensor processed by one lane
+ *
+ * This function computes these offsets for axies independently
+ *
+ * @param rewriter
+ * @param loc
+ * @param elemsPerInstr operand tile shape consumed by one WMMA instruction
+ * @param warpId id component of 2d warp grid along nono-K axis
+ * @param laneId lane id in warp [0..63]
+ * @param warpsPerGroup number of warps in one block
+ * @param numOfElems number of elements accessed by thread per repetition
+ * @param reps number of instructions repretition to fully cover dot operand
+ * @param smemStrides strides in LDS tensor
+ * @param loadVecSize number of elements loaded by one operation
+ * @param iNonKDim non-K dimension of dot operand
+ * @return vector (i-th element corresponds to i-th load instruction) of
+ * 2-element vectors(tensor row and col).
+ */
+llvm::SmallVector<llvm::SmallVector<Value>>
+computeTensorElemMapping(ConversionPatternRewriter &rewriter, Location loc,
+                         const ArrayRef<int64_t> &elemsPerInstr, Value warpId,
+                         Value laneId, int warpsPerGroup, int numOfElems,
+                         ArrayRef<int64_t> reps, ArrayRef<Value> smemOffsets,
+                         int loadVecSize, unsigned iNonKDim,
+                         [[maybe_unused]] unsigned iKDim) {
+  auto numM = reps[0];
+  auto numK = reps[1];
+  const int loadsPerThread = numOfElems / loadVecSize;
+  llvm::SmallVector<llvm::SmallVector<Value>> mapping(numM * numK *
+                                                      loadsPerThread);
+
+  Value _0 = i32_val(0);
+  Value nonKDim = i32_val(iNonKDim);
+
+  for (int block = 0; block < numM; ++block) {
+    Value blockVOffset = i32_val(block * elemsPerInstr[0] * warpsPerGroup);
+    Value blockHOffset = _0;
+    Value warpVOffset = mul(warpId, i32_val(elemsPerInstr[0]));
+    Value warpHOffset = _0;
```

Comment:
May I know why the warpHOffset and blockHOffset are always 0 ? Does this mean that we always have blocks/warps in the vertical dim ?

Reply:
Yes. This is because each warp/block always handle full rows of operand A or full columns of operand B.